rust/otel-arrow-rust/src/encode/producer.rs

```diff
@@ -57,11 +54,61 @@ impl StreamProducer {
     }
 }
 
+const PAYLOAD_TYPE_COUNT: usize = 28; // 28 variants total
+
+// Compile-time lookup table for O(1) conversion
+const PAYLOAD_TYPE_TO_INDEX: [Option<u8>; 46] = {
+    let mut table = [None; 46];
+    table[0] = Some(0); // Unknown
+    table[1] = Some(1); // ResourceAttrs
+    table[2] = Some(2); // ScopeAttrs
+    table[10] = Some(3); // UnivariateMetrics
+    table[11] = Some(4); // NumberDataPoints
+    table[12] = Some(5); // SummaryDataPoints
+    table[13] = Some(6); // HistogramDataPoints
+    table[14] = Some(7); // ExpHistogramDataPoints
+    table[15] = Some(8); // NumberDpAttrs
+    table[16] = Some(9); // SummaryDpAttrs
+    table[17] = Some(10); // HistogramDpAttrs
+    table[18] = Some(11); // ExpHistogramDpAttrs
+    table[19] = Some(12); // NumberDpExemplars
+    table[20] = Some(13); // HistogramDpExemplars
+    table[21] = Some(14); // ExpHistogramDpExemplars
+    table[22] = Some(15); // NumberDpExemplarAttrs
+    table[23] = Some(16); // HistogramDpExemplarAttrs
+    table[24] = Some(17); // ExpHistogramDpExemplarAttrs
+    table[25] = Some(18); // MultivariateMetrics
+    table[26] = Some(19); // MetricAttrs
+    table[30] = Some(20); // Logs
+    table[31] = Some(21); // LogAttrs
+    table[40] = Some(22); // Spans
+    table[41] = Some(23); // SpanAttrs
+    table[42] = Some(24); // SpanEvents
+    table[43] = Some(25); // SpanLinks
+    table[44] = Some(26); // SpanEventAttrs
+    table[45] = Some(27); // SpanLinkAttrs
+    table
+};
+
+impl ArrowPayloadType {
+    #[inline]
+    const fn to_index(self) -> usize {
+        // Safety: Generated enum values are always mapped in the `PAYLOAD_TYPE_TO_INDEX` table.
+        PAYLOAD_TYPE_TO_INDEX[self as usize].expect("`PAYLOAD_TYPE_TO_INDEX` should cover all ArrowPayloadType variants") as usize
+    }
+}
+
+/// Storage for a stream producer with its schema ID
+struct ProducerEntry {
+    schema_id: String,
+    producer: StreamProducer,
+}
+
 /// Produces OTAP `BatchArrowRecords` from OTAP Batches
 pub struct Producer {
     next_batch_id: i64,
     next_schema_id: i64,
-    stream_producers: HashMap<String, StreamProducer>,
+    stream_producers: [Option<ProducerEntry>; PAYLOAD_TYPE_COUNT],
```

Comment:
The perf benefit of having `O(1)` access does come with increased memory usage. For reference, the size of this array is around `5.4KB`. We could reduce it further by having dedicated `Producers` for each signal. Currently, we use the same producer for all signals which forces us to keep 28 items in the array.